File: it-agents/skills/code-architect/references/code-examples/cpp/Observer.cpp
// Observer — subscription mechanism to notify subscribers of events.
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

class Observer {
public:
    virtual ~Observer() = default;
    virtual void onNotify(int value) = 0;
};

class Subject {
    std::vector<Observer*> observers;
    int value = 0;
public:
    void subscribe(Observer* o) { observers.push_back(o); }
    void unsubscribe(Observer* o) {
        observers.erase(std::remove(observers.begin(), observers.end(), o), observers.end());
    }
    void setValue(int v) {
        value = v;
        for (auto* o : observers) o->onNotify(value);
    }
};

class Logger : public Observer {
    std::string name;
public:
    explicit Logger(std::string n) : name(std::move(n)) {}
    void onNotify(int value) override {
        std::cout << name << " observed value=" << value << "\n";
    }
};

int main() {
    Subject subject;
    Logger a{"A"}, b{"B"};
    subject.subscribe(&a);
    subject.subscribe(&b);

    subject.setValue(10);
    subject.unsubscribe(&a);
    subject.setValue(20);
    return 0;
}

File: it-agents/skills/code-architect/references/code-examples/rust/observer.rs
// Observer — define a subscription mechanism so a subject can notify any number
// of dependent observers automatically when its state changes.

use std::cell::RefCell;
use std::rc::Rc;

trait Observer {
    fn on_event(&self, value: i32);
}

struct Logger {
    name: String,
}

impl Observer for Logger {
    fn on_event(&self, value: i32) {
        println!("{} observed value = {}", self.name, value);
    }
}

// Subject maintains a list of subscribers and notifies them on change.
struct Subject {
    observers: Vec<Rc<dyn Observer>>,
    value: i32,
}

impl Subject {
    fn new() -> Self {
        Subject { observers: Vec::new(), value: 0 }
    }

    fn subscribe(&mut self, obs: Rc<dyn Observer>) {
        self.observers.push(obs);
    }

    fn set_value(&mut self, value: i32) {
        self.value = value;
        for obs in &self.observers {
            obs.on_event(value);
        }
    }
}

fn main() {
    let subject = Rc::new(RefCell::new(Subject::new()));
    subject.borrow_mut().subscribe(Rc::new(Logger { name: "A".into() }));
    subject.borrow_mut().subscribe(Rc::new(Logger { name: "B".into() }));

    subject.borrow_mut().set_value(7);
    subject.borrow_mut().set_value(42);
}
